qiime2/sdk/usage.py

```diff
@@ -346,13 +372,37 @@ def _assert_has_line_matching_(self, ref, label, path, expression):
 
 class ExecutionUsage(Usage):
     def _init_data_(self, ref, factory):
-        return factory()
+        result = factory()
+        result_type = type(result)
+
+        if result_type not in (list, set, sdk.Artifact):
+            raise ValueError('Factory (%r) returned a %s, expected an '
+                             'Artifact.' % (factory, result_type))
+
+        if result_type in (list, set):
+            if not all([isinstance(i, sdk.Artifact) for i in result]):
+                raise ValueError('Factory (%r) returned a %s where not all '
+                                 'elements were Artifacts.' %
+                                 (factory, result_type))
+
+        return result
+
+    def _init_metadata_(self, ref, factory):
+        data = factory()
+        if not isinstance(data, Metadata):
+            raise TypeError(
+                f"data must be of type Metadata: type(data) == {type(data)}")
+        return data
```

Comment:
Can you unify the approach used here with the approach above in `_init_data`? They should preferably use the same variable and formatting conventions as above.

Also, please use %-style string formatting, for consistency. Thanks!